### PR TITLE
Automatic Jazzy Docs generation

### PR DESCRIPTION
**Pull request by ShihabMehboob**
This is for [#514](https://github.ibm.com/arf/planning-swift-solutions/issues/514) to automate the generation of Jazzy Docs, which can be triggered by passing in the flagged word `Docs` anywhere in the commit message.

### CODE REVIEW

**Review by rolivieri — CHANGES_REQUESTED** [review]
@ShihabMehboob Thanks for the PR. Can we make a few changes to it? I am envisioning something along these lines:

1) Update `build-package.sh`. For instance, we could add code (after this line https://github.com/IBM-Swift/Package-Builder/blob/521ea64cc5c14679ba1c0adc104ee7b4d9755724/build-package.sh#L166) that is specific to jazzy... something along these lines:

```
# Generate jazzy docs (macOS)
if [ "$(uname)" == "Darwin" ]; then
	invoke script for jazzy doc generation
```
2) We implement a new bash script that contains the logic for jazzy doc generation... we could call this file `jazzy.sh`. In this file, we would have logic that:

2.1) Verifies that a `.jazzy.yaml` does exist in the root folder of the repo. For instance, see https://github.com/IBM-Swift/CloudEnvironment/blob/master/.jazzy.yaml.
2.2) If `.jazzy.yaml` does **not** exist, then we do not attempt to generate jazzy docs.
2.3) If `.jazzy.yaml` does exist, then we attempt to generate the docs:
	- install jazzy
	- generate xcode project
	- run jazzy

Now, regarding the command for generating xcode project... we probably want to leverage the code we already have at: https://github.com/IBM-Swift/Package-Builder/blob/1294bef1ffd89d74536ce06fab97a1c0c4919a37/codecov.sh#L35. Hence, this implies a little bit of refactoring of the existing code. 

2.4) Once we have generated the jazzy docs, we should then commit those changes and push them to the corresponding branch of the repo. Note that we don't want to create an infinite build loop, which requires us to use a git command that looks like this `git commit -m 'documentation update [ci skip]'`. @ddunn2 can give you more details on this.

File: .travis.yml [hunk 1]
```diff
@@ -10,6 +10,7 @@ branches:
   only:
     - master
     - develop
+    - jazzy_travis
```
Comment (rolivieri) [inline .travis.yml]: I am thinking we don't need this line. Can we remove it? Also, you probably may want to rename your branch to `issue.jazzy_travis`.

File: .travis.yml [hunk 2]
```diff
@@ -49,6 +50,12 @@ matrix:
       sudo: required
       env: SWIFT_SNAPSHOT=swift-3.1-DEVELOPMENT-SNAPSHOT-2017-06-14-a
 
+before_install:
```
Comment (rolivieri) [inline .travis.yml]: I don't think this whole section (`before_install`) is needed. Can it be removed?

File: .travis.yml [hunk 3]
```diff
@@ -49,6 +50,12 @@ matrix:
...
+  - sudo apt-get -qq update
+  - sudo apt-get -y -qq install clang lldb-3.8 libicu-dev libtool libcurl4-openssl-dev libbsd-dev build-essential libssl-dev uuid-dev tzdata libz-dev
+  - git clone https://github.com/IBM-Swift/Package-Builder.git
+  - if [[ "$TRAVIS_OS_NAME" == "osx" && $TRAVIS_COMMIT_MESSAGE == *"Docs"* ]]; then ./jazzy.sh; fi
+
```
Comment (rolivieri) [inline .travis.yml]: For this PR, I don't think we need any major changes to the travis.yml file.

File: jazzy.sh [hunk 4]
```diff
@@ -0,0 +1,53 @@
+osName="osx"
```
Comment (rolivieri) [inline jazzy.sh]: We should update the logic in this file (`jazzy.sh`). Looking at the current contents of `jazzy.sh`, there is quite a bit code that I don't think is needed. For instance, I don't think lines 1 - 37 are needed. The Swift binaries are already installed by the time we invoke `jazzy.sh`. What we need instead (in `jazzy.sh`) is something along these lines:

1) Verify that a `.jazzy.yaml` does exist in the root folder of the repo. For instance, see https://github.com/IBM-Swift/CloudEnvironment/blob/master/.jazzy.yaml.
2) If `.jazzy.yaml` does **not** exist, then we do not attempt to generate jazzy docs.
3) If `.jazzy.yaml` **does** exist, then we attempt to generate the docs:
- Install jazzy
- Generate xcode project (see comments below)
- Run jazzy
- Once we have generated the jazzy docs, we should then commit those changes and push them to the corresponding branch of the repo. Note that we don't want to create an infinite build loop, which requires us to use a git command that looks like this git commit -m 'documentation update [ci skip]'. @ddunn2 can give you more details on this.

Now, regarding the logic for generating xcode project, we should leverage the code we already have at https://github.com/IBM-Swift/Package-Builder/blob/1294bef1ffd89d74536ce06fab97a1c0c4919a37/codecov.sh#L35. However, let's avoid code duplication (aka copy and paste). :-) In order to leverage that code, we should do a little bit of refactoring of the existing code found in `codecov.sh`. This way we can use the same code for generating the xcode project from both places, `codecov.sh` and `jazzy.sh`. This implies extracting the Xcode project code generation to a separate file so we can then call it from `codecov.sh` and `jazzy.sh`.

**Review by rolivieri — CHANGES_REQUESTED** [review]
@ShihabMehboob I added more inline comments. Hope that helps. If you'd like to chat on the phone, let's certainly do that, no problem, thanks!

**Comment by rolivieri** [comment]
We should also update the automated test cases to validate/execute this new functionality.

File: generate-xcodeproj.sh [hunk 5]
```diff
@@ -0,0 +1,30 @@
+##
+# Copyright IBM Corporation 2016,2017
```
Comment (sandmman) [inline generate-xcodeproj.sh]: We should update the copyright headers for any of the files we update. Just tack on 2018

**Review by sandmman — CHANGES_REQUESTED** [review]
Just looking this over we are going to need to address a couple of things.

1. It seems as though there is a scoping problem - https://travis-ci.org/IBM-Swift/Package-Builder/jobs/334759276#L200
2. We only want to run this when we push to master, so we should have the same guard-esq statement that codecov uses https://github.com/ShihabMehboob/Package-Builder/blob/0fd5f243a5e75db27b2663e1d73808aa5b259884/codecov.sh#L19. If there’s a logical place you find, this could be extracted and only called once
3. You should be able to add a test case that checks that the jazzy docs were created for one of the example projects we have.
4. I’m not sure what happens when the GitHub credentials aren’t found or fail, but we should check that since it could cause all of our builds to break
5. Copyrights headers

Let me know if you have any questions/comments!

**Comment by ShihabMehboob** [comment]
@sandmman Regarding points 3 and 4:

- A test case in Package-builder for this or in a repo that uses it, like Kitura Contracts?
- Checks currently exist for whether a jazzy.yaml file exists and whether it's on master, would this not imply GitHub credentials existing?

**Comment by sandmman** [comment]
@ShihabMehboob 

1. In [./tests](https://github.com/IBM-Swift/Package-Builder/tree/master/tests), there are two example projects. We could add a .jazzy.yml file to each, then after we execute the build script you can just check that the jazzy files were created.

2. This would imply that the credentials exist for repos that already have logic to push to branch and only imply that for IBM@Swift repos. The majority of the IBM@Swift repos build on master and have a .jazzy.yaml file, but very few have credentials available. As I mentioned previously, if this doesn't actually throw an error causing a failure on the travis build it might not be a big deal, but we should at least have logging if we're not going to introduce some sort of required parameter.

For instance, KituraContracts has a .jazzy yaml file and we build on pushes to master, but it doesn't have any credentials because there was never a need to push to a branch.

**Comment by ShihabMehboob** [comment]
@sandmman This has been tested, and regarding the credentials check and logs, this exists in the build-package.sh file where the directory is checked for existing credentials and appropriately echoed if none exists.

**Comment by sandmman** [comment]
@ShihabMehboob 
1. The jazzy files don't seem to be created in travis. I'm still seeing the same [error](https://travis-ci.org/IBM-Swift/Package-Builder/jobs/340430335#L200) I pointed out the other day.
2. Even though the testing credentials are checked at the start of the script they do not stop the jazzy logic from being executed.

Running the below without credentials returns an error 1, which will fail the entire travis build. This cannot happen. In some way, the git commands must be protected by a credentials check.
```
git add .
git commit -am "Jazzy"
git push origin master

```

**Comment by ShihabMehboob** [comment]
@sandmman The checks have been added which prevents the jazzy generation running if credentials don't exist.

**Review by sandmman — APPROVED** [review]

**Pull request merged**